src/type_icons.c

```diff
@@ -0,0 +1,508 @@
+{
+    struct Sprite* sprite;
+    const struct SpriteTemplate* spriteTemplate = gTypesInfo[type].useSecondTypeIconPalette ? &sSpriteTemplate_TypeIcons2 : &sSpriteTemplate_TypeIcons1;
+    u32 spriteId = CreateSpriteAtEnd(spriteTemplate, x, y, UCHAR_MAX);
+
+    if (spriteId == MAX_SPRITES)
+        return;
+
+    sprite = &gSprites[spriteId];
+    sprite->tMonPosition = position;
+    sprite->tBattlerId = battler;
+    sprite->tVerticalPosition = y;
+
+    sprite->hFlip = ShouldFlipTypeIcon(useDoubleBattleCoords, position, type);
+
+    StartSpriteAnim(sprite, type);
+}
+
+static bool32 ShouldFlipTypeIcon(bool32 useDoubleBattleCoords, u32 position, u32 typeId)
+{
+    bool32 side = (useDoubleBattleCoords) ? B_SIDE_OPPONENT : B_SIDE_PLAYER;
+
+    if (GetBattlerSide(GetBattlerAtPosition(position)) != side)
+        return FALSE;
+
+    return !gTypesInfo[typeId].isSpecialCaseType;
+}
+
+static void SpriteCB_TypeIcon(struct Sprite* sprite)
+{
+    u32 position = sprite->tMonPosition;
+    u32 battlerId = sprite->tBattlerId;
+    bool32 useDoubleBattleCoords = UseDoubleBattleCoords(GetBattlerAtPosition(position));
+
+    if (sprite->tHideIconTimer == NUM_FRAMES_HIDE_TYPE_ICON)
+    {
+        DestroyTypeIcon(sprite);
+        return;
+    }
+
+    if (ShouldHideTypeIcon(battlerId))
+    {
+        sprite->x += GetTypeIconHideMovement(useDoubleBattleCoords, position);
+        ++sprite->tHideIconTimer;
+        return;
+    }
+
+    sprite->x += GetTypeIconSlideMovement(useDoubleBattleCoords,position, sprite->x);
+    sprite->y = GetTypeIconBounceMovement(sprite->tVerticalPosition,position);
+}
+
+static void DestroyTypeIcon(struct Sprite* sprite)
+{
+    u32 i;
+    DestroySpriteAndFreeResources(sprite);
+
+    for (i = 0; i < MAX_SPRITES; ++i)
+    {
+        if (!gSprites[i].inUse)
+            continue;
+
+        if (gSprites[i].template->paletteTag == TYPE_ICON_TAG)
+            return;
+    }
+
+    FreeSpritePaletteByTag(TYPE_ICON_TAG);
+    FreeSpritePaletteByTag(TYPE_ICON_TAG_2);
+}
+
+static bool32 ShouldHideTypeIcon(u32 battlerId)
+{
+    return gBattlerControllerFuncs[battlerId] != PlayerHandleChooseMove
+        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove
+        && gBattlerControllerFuncs[battlerId] != HandleChooseMoveAfterDma3
+        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove
+        && gBattlerControllerFuncs[battlerId] != HandleInputChooseTarget
+        && gBattlerControllerFuncs[battlerId] != HandleMoveSwitching
+        && gBattlerControllerFuncs[battlerId] != HandleInputChooseMove;
+}
+
+static s32 GetTypeIconHideMovement(bool32 useDoubleBattleCoords, u32 position)
+{
+    return useDoubleBattleCoords ?
+        (position == B_POSITION_PLAYER_LEFT ||
+         position == B_POSITION_PLAYER_RIGHT ? 1 : -1) :
+        (position == B_POSITION_PLAYER_LEFT ? -1 : 1);
```

Comment:
Could this be just a normal if? It is really hard to read.

Reply:
https://github.com/rh-hideout/pokeemerald-expansion/pull/5131/commits/654844b9321a12eb1ebd1c918c3d5e86eca028fd